wfe2/verify.go

```diff
@@ -292,11 +294,17 @@ func (wfe *WebFrontEndImpl) matchJWSURLs(outer, inner *jose.JSONWebSignature) *p
 	return nil
 }
 
+// bJSONWebSignature wraps a *jose.JSONWebSignature, but ensures it has been
+// created and processed by the various safety/sanity checks in wfe.parseJWS.
+type bJSONWebSignature struct {
+	*jose.JSONWebSignature
+}
+
 // parseJWS extracts a JSONWebSignature from a byte slice. If there is an error
 // reading the JWS or it is unacceptable (e.g. too many/too few signatures,
-// presence of unprotected headers) a problem is returned, otherwise the parsed
-// *JSONWebSignature is returned.
-func (wfe *WebFrontEndImpl) parseJWS(body []byte) (*jose.JSONWebSignature, *probs.ProblemDetails) {
+// presence of unprotected headers) a problem is returned, otherwise a
+// bJSONWebSignature is returned.
+func (wfe *WebFrontEndImpl) parseJWS(body []byte) (bJSONWebSignature, *probs.ProblemDetails) {
```

Comment:
Because bJSONWebSignature wraps a pointer, we should also pass it by reference. Part of the reason for this is that we want to maintain the invariant that "a bJSONWebSignature always has a Signatures slice of length 1", and here we are creating a bunch of empty bJSONWebSignatures with no Signatures slice at all!

```suggestion
func (wfe *WebFrontEndImpl) parseJWS(body []byte) (*bJSONWebSignature, *probs.ProblemDetails) {
```